### ISSUE TITLE
Fatal error when using composer

### ISSUE THREAD

**Issue by entr** [TGMPA/TGM-Plugin-Activation#236]
I'm getting `Fatal error: Call to undefined function __() in ...\vendor\tgm\plugin-activation\class-tgm-plugin-activation.php on line 161` when using like composer package. What do i miss?

**Comment by chetzof** [comment]
I got this error when included autoloader from wp-config.php
However, when i moved the inclusion into theme's functions.php  - error has been fixed.

But obviously would be great to get this working out of the box.

**Comment by entr** [comment]
This makes sense. Not sure if theme's functions.php is the best place to run autoloaded tough. I would hook to `wp_loaded` or `init` and autoload there.

**Comment by GaryJones** [comment]
I guess at the point of wp-config.php, the `__()` and some other WP functions have not yet been defined.

Pinging @Rarst for advice.

**Comment by Rarst** [comment]
Top of `wp-config.php` is common place at the moment, which I recommend myself. Autoload should not do be made to do anything that would break if loaded before WP core in this or other way.

**Comment by GaryJones** [comment]
So the autoloading is pulling in `class-tgm-plugin-activation.php`, but that's just defining classes and functions. What's being called to trigger the error?

**Comment by marcinant** [comment]
:+1:  bump. Is there any workaround?

**Comment by jrfnl** [comment]
@marcinant We need more information to answer this. Please answer the questions asked above and preferably provide us with a [backtrace of the PHP error](https://gist.github.com/jrfnl/5925642).

**Comment by GaryJones** [comment]
Closing due to lack of feedback.

**Comment by renegeuze** [comment]
The problem remains and I'm happy to give some explanation.

The class-tgm-plugin-activation really did just define classes and functions like  `did_action` on #L1915
That works great as long as you include the autoloader somewhere after the big initialisation of Wordpress functions but does not work when including it earlier.

That concludes the part of what you already know and maybe even intended.

When you setup Wordpress with for example, [Bedrock](https://roots.io/bedrock/), and put tgmpa in the root composer file it will be loaded too early and you'll have no control over it.

Unfortunately I have no ready to go solution that doesn't break backwards compatibility except for skipping tgmpa and use wppackagist for plugins without tgmpa. But you'd lose the nice features tgmpa has.

Example error trace:

	( ! ) Fatal error: Call to undefined function did_action() in [...]/vendor/tgmpa/tgm-plugin-activation/class-tgm-plugin-activation.php on line 1910
	Call Stack
	#	Time	Memory	Function	Location
	1	0.0016	363712	{main}( )	../themes.php:0
	2	0.0023	423464	require_once( '[...]/web/wp/wp-admin/admin.php' )	../themes.php:10
	3	0.0025	438800	require_once( '[...]/web/wp/wp-load.php' )	../admin.php:31
	4	0.0026	441576	require_once( '[...]/web/wp-config.php' )	../wp-load.php:42
	5	0.0027	443264	require_once( '[...]/vendor/autoload.php' )	../wp-config.php:7
	6	0.0029	455472	ComposerAutoloaderInitcfa8913385ecad8212dee8e868c2715a::getLoader( )	../autoload.php:7
	7	0.0039	527792	composerRequirecfa8913385ecad8212dee8e868c2715a( )	../autoload_real.php:45
	8	0.0097	1106488	require( '[...]/vendor/tgmpa/tgm-plugin-activation/class-tgm-plugin-activation.php' )	../autoload_real.php:54

**Comment by Rarst** [comment]
Yeah, this is "wrong" to some degree. Also see #488 

#265 if done would make it go away.

**Comment by jrfnl** [comment]
I'll try to have a look at this in more detail next week. Sorry, can't do much at this moment #zendcon.

**Comment by renegeuze** [comment]
I just realised that it would not break backwards compatibility if all those parts were simply wrapped inside function_exists(best watch for 1 single function/constant that's defined when ready).

Of course the nicer way would be to go the PSR-4 way(bump major version?) but just some scenarios using the function_exist/is_defined:

1. Require tgmpa-class from inside theme functions.php - no problem before, no problem after.
2. Require tgmpa-class from inside config.php(or elsewhere) - didn't work before anyway so people did not do that, still doesn't work afterwards
3. Autoloader from inside theme functions.php - Works before, works after.
4. Autoloader from inside config.php(or anywhere) - Broken before,  safe to do after but you still need to add the file as per scenario 1 as require and not require_once.

**Comment by tlartaud** [comment]
Any news on this ? Still needing .. Still broken :( I can't believe a so huge issue is still open. Is TGMPA a discontinued project ? I can still use my Fork where I simply removed the autoload, but how about your suggestions ? function_exists and so on ? Did someone try to fix it ? @jrfnl ? @Rarst ? @renegeuze ?
This PHP framework is using TGMPA : https://github.com/wp-jungle/baobab
I had to change the main repo to my fork one because this framework is also intented to be loaded into a project root (not into a WP Theme). Obviously, it won't work with the actual autoload in a project root. Sometimes, we have to accept some breaking changes and start fixing the things instead of saying : "It's broken till the beginning, so we keep it broken in case that someone use that broken thing". Completely senseless.

**Comment by jrfnl** [comment]
> I can't believe a so huge issue is still open

This really is not a huge issue. Not that many people in the WP community using Composer in that way. TGMPA is intended to be shipped with themes/plugins and those should be complete packages with any dependencies already included. They might use Composer for dev, but not for deployment.
When used like that, there is no issue.

> Is TGMPA a discontinued project ?

No, it's not, but for the next step of the project to be worth my time (v3), I, for one, want to see interest in getting the basis of TGMPA into WordPress core.
Until that happens, you'll only see minor updates & bug fixes and as we currently rarely get any bug reports - especially since the release of the Custom TGMPA Generator - there is just not much which needs to be done.

> but how about your suggestions ? function_exists and so on ?

That just wouldn't work. It would effectively disable TGMPA if the file is loaded too early.
The onus of loading the file at the right point in the WP load flow is on you as a developer. It's kind of senseless blaming us for you doing it wrong....

**Comment by Rarst** [comment]
> This really is not a huge issue. Not that many people in the WP community using Composer in that way. 

WP core Composer package is now installed upwards of two thousands times a day. I know you don't care for Composer personally, but please don't let that color your judgments about Composer use in WP and demand for it. By now it is significant and only keeps growing.

> The onus of loading the file at the right point in the WP load flow is on you as a developer. It's kind of senseless blaming us for you doing it wrong....

No, TGMPA is doing it wrong by using Composer autoload mode that is incompatible with site–level stacks.  It is not something downstream developer chooses or has control over. You can consider this by design, but responsibility for this choice and implications is firmly on the project, not downstream consumers.

**Comment by jrfnl** [comment]
> TGMPA is doing it wrong by using Composer autoload mode that is incompatible with site–level stacks. 

I have no clue what you are talking about, but feel free to PR a solution if you know of one.

**Comment by Rarst** [comment]
One solution would be #265 (or otherwise untangling definitions from logic), as I wrote above.

**Comment by renegeuze** [comment]
Good news. If tgmpa doesn't work for you, you can just try something else ;-)
For our own projects we're able to install most plugins as Must-Use directly from composer using wpackagist. By shuffling the order of install paths plugins are installed as MU unless otherwise specified.
For plugins that are not capable of running as MU we just force activate them elsewhere.
We have the luxury of having no need for Optional plugins in our tailor made projects.

As part of a real solution, I didn't believe it at first but Wordpress actually has a way of setting subscribers(actions) /before/ the add_action function exists.

The below code could be made part of a function that gets autoloaded using composer and maybe extend something like the Symfony eventDispatcher component.

```
// Function allows adding wp hooks before wp is initialized
function aym_add_action(
    string $tag,
    $function,
    int $priority = 10,
    int $accepted_args = 1
) {
    global $wp_filter;

    if (!is_array($wp_filter)) {
        $wp_filter = [];
    }

    if (!is_array($wp_filter[$tag])) {
        $wp_filter[$tag] = [];
    }

    if (!is_array($wp_filter[$tag][$priority])) {
        $wp_filter[$tag][$priority] = [];
    }

    $wp_filter[$tag][$priority][] = [
        'function' => $function,
        'accepted_args' => $accepted_args,
    ];
};
```

```
// Example hooking into WP init
aym_add_action('init', function() {
    echo 'Early bird';
});
```

And definitely +1 on #265

**Comment by mindfullsilence** [comment]
As a workaround, you can redefine the package yourself in your projects composer.json like so, emptying the autoload["files"] key:

```json
{
  ...
  "repositories": [

    {
      "type": "package",
      "package": {
        "name": "tgmpa/tgm-plugin-activation",
        "version": "2.6.1",
        "type": "package",
        "dist": {
          "url": "https://github.com/TGMPA/TGM-Plugin-Activation/archive/2.6.1.zip",
          "type": "zip"
        },
        "autoload": {
          "files": []
        }
      },
      "url": "https://github.com/TGMPA/TGM-Plugin-Activation.git"
    },
    ...
  ],
  "require": {
    ...
    "tgmpa/tgm-plugin-activation": "*"
  }
}
```

**Comment by jrfnl** [comment]
@mindfullsilence We made a change quite a while ago (not in a released version yet), which should prevent these type of issues even when using the composer file which is included with TGMPA. See #665
Would be great if you test with `master` to see if that change solves your issue too.

**Comment by lbngoc** [comment]
I use this command at this time to get current branch with this fix #594 `composer require  tgmpa/tgm-plugin-activation:dev-develop`
